### PR TITLE
Fixed time update, Bifrost stage update, and USD stage file update

### PR DESCRIPTION
**Pull request by ppt-adsk**
With tests for time update and USD stage file update.

### CODE REVIEW

File: lib/flowViewport/fvpInstruments.h
Comment (ppt-adsk) [inline lib/flowViewport/fvpInstruments.h:1]: Added trivial string keyed dictionary to keep track of instrumentation values in mayaHydra.  Useful for reasoning and assertions.

File: test/testSamples/testUsdAnim/cubeAnim.usda
Comment (ppt-adsk) [inline test/testSamples/testUsdAnim/cubeAnim.usda:1]: USD cube with animated translation.

File: lib/mayaHydra/hydraExtensions/sceneIndex/registration.cpp
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/registration.cpp:207]: This is the central construct in fixing HYDRA-930: we receive notice of time invalidation and immediately call compute() on the proxy shape output time, during dirty propagation.  This returns the previous time, which is incorrect.  Instead, simply refresh time for the USD data producer at render time, at which point the proxy shape output compute() produces the correct result.  This matches the way the VP 2 render delegate sets time.

File: lib/mayaHydra/hydraExtensions/sceneIndex/registration.h
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/registration.h:63]: New hook method to upate the registered data provider at render time.  Used by USD registration to update time.

File: lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.cpp
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.cpp]: Apologies, no real change here, debugging only, will revert when I do review fixes.

File: lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h:39]: Updating time at every render, no need to be listening to time changes.

File: lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h:46]: Class MayaUsdProxyShapeSceneIndex was both in the Pixar namespace and the mayaHydra namespace, which was a bit much, belongs in the mayaHydra namespace only.

File: lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h:67]: Key for number of populate calls instrument.

File: lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h
Comment (ppt-adsk) [inline lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.h:104]: Now listening for proxy shape stage invalidate notifications, which include integer stage ID increments, USD file path changes, etc.  See
https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/nodes/proxyShapeBase.cpp

File: lib/mayaHydra/mayaPlugin/renderOverride.cpp
Comment (ppt-adsk) [inline lib/mayaHydra/mayaPlugin/renderOverride.cpp:1178]: Allow registered plugins to update before render.  Our USD registered data provider uses this to update time.

File: test/lib/mayaUsd/render/mayaToHydra/CMakeLists.txt
Comment (ppt-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/CMakeLists.txt:80]: Discovered that time computation (pulling on the proxy shape time output) was being done by the VP 2 render delegate, which meant that this value was being provided to the Maya Hydra viewport as a side effect.  Make sure that tests run without the VP 2 render delegate to insulate the test from this.  We should revisit this and perhaps turn of VP 2 render delegate updates when it is not the viewport renderer, while ensuring that correctness is maintained.

File: test/lib/mayaUsd/render/mayaToHydra/cpp/testHydraPrim.cpp
Comment (ppt-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/cpp/testHydraPrim.cpp:44]: This is just testing that an application path object has a corresponding Hydra scene index prim.  This could be better done with a Python binding to a C++ function, rather than abusing our test framework test plugin command to do so.

File: test/lib/mayaUsd/render/mayaToHydra/cpp/mayaHydraCppTestsCmd.h
Comment (ppt-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/cpp/mayaHydraCppTestsCmd.h:29]: Query the instruments dictionary from a Python test.  Would be much better done with a Python binding to a C++ function.

File: test/lib/mayaUsd/render/mayaToHydra/cpp/testUsdStageFromFile.py
Comment (ppt-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/cpp/testUsdStageFromFile.py:1]: Test that swapping USD files changes the Hydra scene index scene.

File: test/lib/mayaUsd/render/mayaToHydra/cpp/testUsdStageFromFile.py
Comment (ppt-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/cpp/testUsdStageFromFile.py:56]: Assert that only one populate call was done to replace the USD scene.

File: test/lib/mayaUsd/render/mayaToHydra/cpp/mayaHydraCppTestsCmd.cpp
Comment (ppt-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/cpp/mayaHydraCppTestsCmd.cpp:167]: Query the instruments dictionary through a Maya command, for use in Python tests.  Would be better done through a Python binding of a C++ function.

File: test/lib/mayaUsd/render/mayaToHydra/cpp/testUsdAnim.py
Comment (ppt-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/cpp/testUsdAnim.py:1]: Simple Python driver for the real time varying test in testUsdAnim.cpp.

File: test/lib/mayaUsd/render/mayaToHydra/cpp/testUsdAnim.cpp
Comment (ppt-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/cpp/testUsdAnim.cpp:46]: Run through the animation and check that the z translation value changes correctly.

File: lib/flowViewport/CMakeLists.txt
Comment (debloip-adsk) [inline lib/flowViewport/CMakeLists.txt:19]: I'm assuming we're removing it from the list so we don't ship it, as it's only used for a workaround?

**Review by debloip-adsk — DISMISSED** [review]
Looks good, thanks! I've got two higher-level questions though :
1) Currently, the instruments utility is always used/run, even in release builds; is there not a risk that if we use it too much or for some expensive instrumentation, that it starts negatively affecting performance? Should we consider modifying it to only run in debug builds? This would cause problems for the unit test that depends on it though.
2) For the Python bindings, should we not do them now, or if we need to look into it, log a ticket to that effect?

File: lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.cpp
Comment (debloip-adsk) [reply lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.cpp]: I approved for now since I have no hard changes to request and this is still functional, I'll re-approve once updated

File: lib/flowViewport/CMakeLists.txt
Comment (ppt-adsk) [reply lib/flowViewport/CMakeLists.txt:19]: Good catch!  Purely copy-paste damage, reverted.

File: lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.cpp
Comment (ppt-adsk) [reply lib/mayaHydra/hydraExtensions/sceneIndex/mhMayaUsdProxyShapeSceneIndex.cpp]: Reverted debugging change.

**Comment by ppt-adsk** [comment]
> Looks good, thanks! I've got two higher-level questions though :
> 
>     1. Currently, the instruments utility is always used/run, even in release builds; is there not a risk that if we use it too much or for some expensive instrumentation, that it starts negatively affecting performance? Should we consider modifying it to only run in debug builds? This would cause problems for the unit test that depends on it though.
> 
>     2. For the Python bindings, should we not do them now, or if we need to look into it, log a ticket to that effect?

1. There is a risk, but I think it's manageable, and we can assess on a case by case basis.  I think the gain in understanding the release build behavior outweighs the risk of poor use.
2. I would log a ticket, and wait until Pixar switches to pybind11 for their Python bindings.  This would allow us to use pybind11 across UFE (already using pybind11), mayaHydra, and USD bindings, and provide for easy interchange.  For now, using a plugin Maya command is an acceptable workaround, in my opinion.  Python binding work entered as HYDRA-989.

**Review by debloip-adsk — APPROVED** [review]

**Pull request merged**
